### PR TITLE
feat: add REST endpoint for SmartDocuments

### PR DESCRIPTION
**Pull request by hsiliev**
Add REST endpoints for SmartDocuments
Add converter functions for REST model

Solves: PZ-2723

### CODE REVIEW

File: scripts/docker-compose/imports/smartdocuments-wiremock/__files/sdapi-structure-response-body.json [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+      "templateGroups": [
+        {
+          "templateGroups": [
+            {
+              "templateGroups": [],
+              "templates": [
+                {
+                  "favorite": false,
+                  "id": "7B7857BB9959470C82974037304E433D",
+                  "name": "Data Test"
```
Comment (edgarvonk) [inline scripts/docker-compose/imports/smartdocuments-wiremock/__files/sdapi-structure-response-body.json:13]: Minor thing but these test template names seem a little arbitrary and weird? I would rather name them something like `smartDocumentsTestTemplate1` and `smartDocumentsTestTemplate2` or something?

File: scripts/docker-compose/imports/smartdocuments-wiremock/__files/sdapi-structure-response-body.json [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+              "id": "6BF6F7B751554C24964C7B39DC62F2BA",
+              "name": "Memo nieuw"
+            },
+            {
+              "favorite": false,
+              "id": "4A615E0E9F754539A7BB969B55CC814A",
+              "name": "Memo oud"
+            },
+            {
+              "favorite": false,
```
Comment (edgarvonk) [inline scripts/docker-compose/imports/smartdocuments-wiremock/__files/sdapi-structure-response-body.json:127]: Maybe ok but do we really need all these templates here for our itests? Usually when you have 2 or some data item you are testing enough and it does not make sense to add any more? I am just a bit worried about the maintenance of test data like this so would rather keep it as small as we can.

File: src/itest/kotlin/nl/lifely/zac/itest/SmartDocumentsTest.kt [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+                    shouldContainJsonKeyValue("$[0].templates[0].name", "Aanvullende informatie nieuw")
+                    shouldContainJsonKeyValue("$[0].templates[1].id", "CDC0FD0F08B545B59E36D610AFFA0B58")
+                    shouldContainJsonKeyValue("$[0].templates[1].name", "Aanvullende informatie oud")
+                    shouldContainJsonKeyValue("$[0].templates[2].id", "40A8482A34F941618B95359BF8246FB8")
+                    shouldContainJsonKeyValue("$[0].templates[2].name", "Advies nieuw")
+                    shouldContainJsonKeyValue("$[0].templates[3].id", "16B8B5A8E18A4BA484302029B51C32B6")
+                    shouldContainJsonKeyValue("$[0].templates[3].name", "Advies oud")
+                    shouldContainJsonKeyValue("$[0].templates[4].id", "53AFB00EE74E4B8AB1EC73CCB3251F8B")
+                    shouldContainJsonKeyValue("$[0].templates[4].name", "Besluit nieuw")
+                    shouldContainJsonKeyValue("$[0].templates[5].id", "60D711D768184F7F94A8BC1553A689A1")
```
Comment (edgarvonk) [inline src/itest/kotlin/nl/lifely/zac/itest/SmartDocumentsTest.kt:76]: Same here. To me testing so many templates here does not add any value?

**Review by edgarvonk — CHANGES_REQUESTED** [review]

File: src/main/kotlin/net/atos/zac/smartdocuments/templates/SmartDocumentsTemplateConverter.kt [hunk 4]
```diff
@@ -3,49 +3,147 @@ package net.atos.zac.smartdocuments.templates
...
         this.documentsStructure.templatesStructure.templateGroups
-            .mapTo(mutableSetOf()) { convertTemplateGroupToModel(it, null) }
+            .mapTo(mutableSetOf()) { convertTemplateGroupResponseToREST(it) }
+
+    private fun convertTemplateGroupResponseToREST(
+        group: SmartDocumentsResponseTemplateGroup
+    ): RESTSmartDocumentsTemplateGroup {
+        val restGroup = createRESTTemplateGroup(group)
+
+        if (!group.templates.isNullOrEmpty()) {
```
Comment (bas-info-nl) [inline src/main/kotlin/net/atos/zac/smartdocuments/templates/SmartDocumentsTemplateConverter.kt:25]: I think this is better achieved like so:
```kotlin
restGroup.templates = group.templates?.map { createRESTTemplate(it) }?.toSet()
```
Or is there a logic error?

File: src/main/kotlin/net/atos/zac/smartdocuments/templates/SmartDocumentsTemplateConverter.kt [hunk 5]
```diff
@@ -3,49 +3,147 @@ package net.atos.zac.smartdocuments.templates
...
+            restGroup.templates = mutableSetOf<RESTSmartDocumentsTemplate>().apply {
+                group.templates?.forEach {
+                    add(createRESTTemplate(it))
+                }
+            }
+        }
+
+        if (!group.templateGroups.isNullOrEmpty()) {
```
Comment (bas-info-nl) [inline src/main/kotlin/net/atos/zac/smartdocuments/templates/SmartDocumentsTemplateConverter.kt:33]: Similarly here...
```kotlin
restGroup.groups = group.templateGroups?.map { convertTemplateGroupResponseToREST(it) }?.toSet()
```

File: src/main/kotlin/net/atos/zac/smartdocuments/templates/SmartDocumentsTemplateConverter.kt [hunk 6]
```diff
@@ -3,49 +3,147 @@ package net.atos.zac.smartdocuments.templates
...
     ) = SmartDocumentsTemplate().apply {
         smartDocumentsId = smartDocumentsTemplate.id
+        zaakafhandelParameters = zaakafhandelParams
         name = smartDocumentsTemplate.name
+        templateGroup = parentGroup
         creationDate = ZonedDateTime.now()
     }
+
+    fun Set<SmartDocumentsTemplateGroup>.toREST(): Set<RESTSmartDocumentsTemplateGroup> =
+        this.mapTo(mutableSetOf()) { convertTemplateGroupToREST(it) }
```
Comment (bas-info-nl) [inline src/main/kotlin/net/atos/zac/smartdocuments/templates/SmartDocumentsTemplateConverter.kt:108]: I am not sure why we need these `mutableSet`s, these are just model classes right? We're not going to make any changes to the sets nor contained data, correct?

File: src/main/kotlin/net/atos/zac/smartdocuments/templates/SmartDocumentsTemplateConverter.kt [hunk 7]
```diff
@@ -3,49 +3,147 @@ package net.atos.zac.smartdocuments.templates
...
+
+    private fun convertTemplateGroupToREST(
+        group: SmartDocumentsTemplateGroup
+    ): RESTSmartDocumentsTemplateGroup {
+        val restTemplateGroup = createRESTTemplateGroup(group)
+
+        if (group.templates.isNotEmpty()) {
+            restTemplateGroup.templates = mutableSetOf<RESTSmartDocumentsTemplate>().apply {
```
Comment (bas-info-nl) [inline src/main/kotlin/net/atos/zac/smartdocuments/templates/SmartDocumentsTemplateConverter.kt:116]: This conversion should also be written in a more functional way, then it'll be a one liner. Like the other suggestions I made, and similarly with the below code.

**Review by bas-info-nl — APPROVED** [review]
Approved with some comments

File: scripts/docker-compose/imports/smartdocuments-wiremock/__files/sdapi-structure-response-body.json (hunk 2 above)
Comment (hsiliev) [reply scripts/docker-compose/imports/smartdocuments-wiremock/__files/sdapi-structure-response-body.json:127]: Left only 2 templates

File: src/itest/kotlin/nl/lifely/zac/itest/SmartDocumentsTest.kt (hunk 3 above)
Comment (hsiliev) [reply src/itest/kotlin/nl/lifely/zac/itest/SmartDocumentsTest.kt:76]: We convert from SD response to our REST model. 
So the mocked response is SD format and the test checks if the REST response has all the fields, as expected.

Comment (hsiliev) [reply src/itest/kotlin/nl/lifely/zac/itest/SmartDocumentsTest.kt:76]: Left only 2 templates to exercise loops & recursion

File: scripts/docker-compose/imports/smartdocuments-wiremock/__files/sdapi-structure-response-body.json (hunk 1 above)
Comment (hsiliev) [reply scripts/docker-compose/imports/smartdocuments-wiremock/__files/sdapi-structure-response-body.json:13]: Renamed to `root`, `group 2 template 1`, etc.

File: src/main/kotlin/net/atos/zac/smartdocuments/templates/SmartDocumentsTemplateConverter.kt (hunk 4 above)
Comment (hsiliev) [reply src/main/kotlin/net/atos/zac/smartdocuments/templates/SmartDocumentsTemplateConverter.kt:25]: Test is ok, which means it's equivalent, so replaced it. Thanks!

File: src/main/kotlin/net/atos/zac/smartdocuments/templates/SmartDocumentsTemplateConverter.kt (hunk 6 above)
Comment (hsiliev) [reply src/main/kotlin/net/atos/zac/smartdocuments/templates/SmartDocumentsTemplateConverter.kt:108]: This comes from the JPA model. ManyToOne and OneToMany cannot be sets, but are required to be mutable

File: src/main/kotlin/net/atos/zac/smartdocuments/templates/SmartDocumentsTemplateConverter.kt (hunk 5 above)
Comment (hsiliev) [reply src/main/kotlin/net/atos/zac/smartdocuments/templates/SmartDocumentsTemplateConverter.kt:33]: Applied

**Review by edgarvonk — APPROVED** [review]
nice!

File: src/main/kotlin/net/atos/zac/smartdocuments/templates/SmartDocumentsTemplateConverter.kt (hunk 6 above)
Comment (hsiliev) [reply src/main/kotlin/net/atos/zac/smartdocuments/templates/SmartDocumentsTemplateConverter.kt:108]: This one in particular is the start of the recursion, so `convertTemplateGroupToREST` adds data to the set.

**Pull request merged**